Content.Server/GameObjects/Components/Observer/GhostComponent.cs

```diff
@@ -62,29 +76,82 @@ public override void HandleMessage(ComponentMessage message, IComponent componen
-                    {
-                        actor.playerSession.ContentData().Mind.UnVisit();
-                        Owner.Delete();
-                    }
+                {
+                    case ReturnToBodyComponentMessage reenter:
+                        if (!Owner.TryGetComponent(out IActorComponent actor) || !CanReturnToBody) break;
+                        if (netChannel == null || netChannel == actor.playerSession.ConnectedClient)
+                        {
+                            actor.playerSession.ContentData().Mind.UnVisit();
+                            Owner.Delete();
+                        }
 
-                    break;
-                case ReturnToCloneComponentMessage reenter:
+                        break;
+                    case ReturnToCloneComponentMessage reenter:
 
-                    if (Owner.TryGetComponent(out VisitingMindComponent mind))
-                    {
-                        Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new GhostReturnMessage(mind.Mind));
-                    }
+                        if (Owner.TryGetComponent(out VisitingMindComponent mind))
+                        {
+                            Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new GhostReturnMessage(mind.Mind));
+                        }
 
-                    break;
-                default:
-                    break;
-            }
+                        break;
+
+                    case GhostWarpRequestMessage warp:
+                        if (warp.PlayerTarget != default)
+                        {
+                            foreach (var player in _playerManager.GetAllPlayers())
+                            {
+                                if (warp.PlayerTarget == player.AttachedEntity.Uid)
+                                {
+                                    session.AttachedEntity.Transform.Coordinates =
+                                        player.AttachedEntity.Transform.Coordinates;
+                                }
+                            }
+                        }
+                        else
+                        {
+                            foreach (var warpPoint in FindWaypoints())
+                            {
+                                if (warp.WarpName == warpPoint.Location)
+                                {
+                                    session.AttachedEntity.Transform.Coordinates = warpPoint.Owner.Transform.Coordinates ;
+                                }
+                            }
+                        }
+                        break;
+
+                    case GhostRequestPlayerNameData data:
+                        var playerNames = new Dictionary<EntityUid, string>();
+                        foreach (var names in _playerManager.GetAllPlayers())
+                        {
+                            if (names.AttachedEntity != null && names.SessionId != netChannel.SessionId)
```

Comment:
Maybe check if the player is alive?

Reply:
Elaborate?

Reply:
If we allow tp'ing to dead players this is okay

Reply:
Definitely allow dead-tp.